### PR TITLE
Add DRY Samplers to ST Staging

### PR DESCRIPTION
**Pull request by kokansei**
Added support for DRY samplers as described in the ooba webui repo here: https://github.com/oobabooga/text-generation-webui/pull/5677

DRY_sampler_SS

Tested and confirmed working with ooba webui but may need to be further tested and revised to work with other backends. I am not a coder by trade and this is my first attempt to contribute to a project on github, so please scrutinize accordingly.

### CODE REVIEW

**Comment by Cohee1207** [comment]
Can you recreate the PR and commit the files without reformatting them?

**Comment by kokansei** [comment]
Okay, I reuploaded the files with the formatting unborked so the diffs look a lot cleaner and github no longer complains about conflicts.

**Comment by kokansei** [comment]
(Cleaned up a few more formatting issues)

File: public/index.html
Comment (Cohee1207) [inline public/index.html]: This only applies to ooba, so leave just ooba in data-tg-type attribute

File: public/index.html
Comment (Cohee1207) [inline public/index.html]: data-tg-type is not needed in nested blocks - it was already hidden by the top-level block. Feel free to remove here (and two other instances below)

**Review by Cohee1207 — CHANGES_REQUESTED** [review]
This is great, but let's wait until TextGen WebUI merges it.

File: public/index.html
Comment (p-e-w) [reply public/index.html]: [An implementation for llama.cpp](https://github.com/ggerganov/llama.cpp/pull/6839) is on the way!

**Comment by p-e-w** [comment]
Great to see this happening! There's one important frontend-specific feature that I had envisioned, namely a checkbox like this (checked by default):

**[X] Add user/character names to sequence breakers**

Sequence breakers are the secret sauce that allows DRY to work where similar approaches have failed. The default list already takes care of chat templates, but exotic character names can sometimes be a problem since the model effectively must spell them out, creating long repeating token sequences. The solution is to add those names (more precisely, the words they are comprised of) to the list of sequence breakers.

But SillyTavern already has that information, so it can perform that step without the user needing to bother. The idea is that if the checkbox is ticked, the names of all chat participants are retrieved, split into words (standard split on space should suffice), and those words are appended to the list of sequence breakers. That way, the user never has to worry about this, even when they switch characters or personalities.

File: public/index.html
Comment (p-e-w) [inline public/index.html]:
```suggestion
                                            <label data-i18n="DRY Repetition Penalty">DRY Repetition Penalty</label>
```

While DRY is of course technically a sampler, this makes it more clear what its purpose is.

File: public/index.html
Comment (p-e-w) [inline public/index.html]:
```suggestion
                                            <span data-i18n="Sequence Breakers">Sequence Breakers</span>
```

DRY is already in the title.

File: public/scripts/textgen-settings.js
Comment (p-e-w) [inline public/scripts/textgen-settings.js]:
```suggestion
	dry_sequence_breakers: '[]',
```

An empty string is not valid JSON. Note that the syntax of this parameter is currently being discussed upstream.

**Comment by p-e-w** [comment]
Note that the llama.cpp implementation of DRY has an additional parameter, `dry_penalty_last_n`, that allows controlling the length of the penalty context independently of the standard repetition penalty. I had originally implemented that in the text-generation-webui PR as well, but the maintainer prefers to not have separate range parameters.

File: public/index.html
Comment (Cohee1207) [reply public/index.html]: Still not merged tho. So we wait again.

File: public/scripts/power-user.js
Comment (DBFrosty) [inline public/scripts/power-user.js]: I think this is causing an error. ") {" => " ||"

Comment (Myobu1) [reply public/scripts/power-user.js]: Black screen error? I cloned the developers staging repo, get a black screen with no interface. Then I tried checking out this PR into a fresh clone of Silly Staging. Navigated to Silly's web UI and get a black screen with no interface. The installation previously worked until I checked out the PR.

Comment (RachidAR) [inline public/scripts/power-user.js]: > I think this is causing an error. ") {" => " ||"

That right. It needs to be fixed.
```suggestion
        sliderID == 'smoothing_factor_textgenerationwebui' ||
```

**Review by RachidAR — CHANGES_REQUESTED** [review]

File: public/index.html
Comment (Cohee1207) [reply public/index.html]: Fixed.

File: public/scripts/power-user.js
Comment (Cohee1207) [reply public/scripts/power-user.js]: Fixed.

Comment (Cohee1207) [reply public/scripts/power-user.js]: Fixed.

**Comment by ike709** [comment]
The WebUI pull request has been merged.

**Comment by p-e-w** [comment]
I have added documentation for each parameter to the text-generation-webui frontend in this commit: https://github.com/oobabooga/text-generation-webui/pull/5677/commits/2585d07e2a0f84f86371b862b80b80daaf76f1d4

It might make sense to copy those documentation strings into this PR, perhaps in the form of tooltips, so that users can find out what the values mean. You could also link to the original PR which has a detailed explanation of how DRY works in general.

**Comment by Cohee1207** [comment]
To be continued when merged for llamacpp. Will create a separate PR for that.

**Review by Cohee1207 — APPROVED** [review]

**Pull request merged**
